File: mcp-server/internal/domains/mcpservers/apply.go
package mcpservers

import (
	"context"
	"fmt"

	mcpserverv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/mcpserver/v1"
	"github.com/stigmer/stigmer/mcp-server/internal/domains"
	"google.golang.org/grpc"
)

// Apply creates or updates an MCP server via the
// McpServerCommandController.Apply RPC. The proto must be fully constructed
// (use McpServerInput.ToProto()).
func Apply(ctx context.Context, serverAddress string, mcpServer *mcpserverv1.McpServer) (string, error) {
	return domains.WithConnection(ctx, serverAddress,
		func(ctx context.Context, conn *grpc.ClientConn) (string, error) {
			client := mcpserverv1.NewMcpServerCommandControllerClient(conn)
			result, err := client.Apply(ctx, mcpServer)
			if err != nil {
				desc := fmt.Sprintf("MCP server %q in org %q", mcpServer.GetMetadata().GetSlug(), mcpServer.GetMetadata().GetOrg())
				return "", domains.RPCError(err, desc)
			}
			return domains.MarshalJSON(result)
		})
}

File: mcp-server-ts/src/domains/mcpservers/apply.ts
// McpServer apply path: create-or-update via the McpServerCommandController.apply
// RPC. The flat MCP input is projected into a fully-formed McpServer proto by the
// generated mcpServerInputToProto bridge (codegen, src/gen/mcpserver.ts), which
// also rebuilds the stdio/http oneof, before the call.
// Go parity: mcp-server/internal/domains/mcpservers/apply.go.

import { McpServerSchema } from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/api_pb";
import { McpServerCommandController } from "@stigmer/protos/ai/stigmer/agentic/mcpserver/v1/command_pb";

import { mcpServerInputToProto, type McpServerInput } from "../../gen/mcpserver";
import { withClient } from "../client";
import { toProtoJson } from "../marshal";
import { rpcError } from "../rpcerr";

/** Create or update an MCP server, returning the persisted server as protojson. */
export async function applyMcpServer(
  serverAddress: string,
  token: string,
  input: McpServerInput,
): Promise<string> {
  const server = mcpServerInputToProto(input);
  const desc = `mcp server "${server.metadata?.slug ?? ""}" in org "${server.metadata?.org ?? ""}"`;
  return withClient(
    McpServerCommandController,
    serverAddress,
    token,
    async (client, callOptions) => {
      try {
        const result = await client.apply(server, callOptions);
        return toProtoJson(McpServerSchema, result);
      } catch (err) {
        throw rpcError(err, desc);
      }
    },
  );
}
